solr/solrj/src/java/org/apache/solr/common/util/CompressionUtil.java

```diff
@@ -0,0 +1,109 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.common.util;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.zip.DataFormatException;
+import java.util.zip.Deflater;
+import java.util.zip.Inflater;
+
+/**
+ * Utility class for dealing with LZ4 compressed data, such as state.json in Zookeeper if configured to be compressed
+ */
+public class CompressionUtil {
+
+    private final static char[] ZLIB_FIRST_MAGIC_HEX = new char[]{'7', '8'};
+    private final static char[] ZLIB_SECOND_MAGIC_HEX = new char[]{'0', '1'};
+    private final static int COMPRESSED_SIZE_MAGIC_NUMBER = 8;
+
+    /**
+     * Uses the hex magic number for zlib compression '78 01' to check if the bytes are compressed
+     * @param data - the bytes to check for compression
+     * @return true if the data is zlib compressed based on the first 2 bytes, false otherwise
+     */
+    public static boolean isCompressedBytes(byte[] data) {
+        if (data == null || data.length < 2) return false;
+        // two characters form the hex value.
+        char[] hexDigits = new char[2];
+        hexDigits[0] = Character.forDigit((data[0] >> 4) & 0xF, 16);
+        hexDigits[1] = Character.forDigit((data[0] & 0xF), 16);
+        if (!Arrays.equals(hexDigits, ZLIB_FIRST_MAGIC_HEX)) return false;
+        hexDigits = new char[2];
+        hexDigits[0] = Character.forDigit((data[1] >> 4) & 0xF, 16);
+        hexDigits[1] = Character.forDigit((data[1] & 0xF), 16);
+        return Arrays.equals(hexDigits, ZLIB_SECOND_MAGIC_HEX);
+    }
+
+    /**
+     * Decompresses zlib compressed bytes, returning the uncompressed data as a byte[]
+     * @param data the input zlib compressed data to decompress
+     * @return the decompressed bytes
+     * @throws DataFormatException - The data is not zlib compressed data
+     */
+    public static byte[] decompressBytes(byte[] data) throws DataFormatException {
+        if (data == null) return null;
+        Inflater inflater = new Inflater();
+        inflater.setInput(data, 0, data.length);
+        // Attempt to get the decompressed size from trailing bytes, this will be present if compressed by Solr
+        int decompressedSize = ByteBuffer.wrap(data, data.length - 8, 4).getInt();
+        int xoredSize = ByteBuffer.wrap(data, data.length - 4, 4).getInt();
+        if ((decompressedSize ^ COMPRESSED_SIZE_MAGIC_NUMBER) != xoredSize) {
+            // Take best guess of decompressed size since it wasn't included in trailing bytes
+            decompressedSize = (int)(1.8*data.length);
```

Comment:
Writing the exact uncompressed length into the trailer should make this case (where we have to guess the size) very uncommon -- that said, if we're picking a ballpark compression ratio to estimate size, I think we'd want to over-estimate, and in any case I'd expect the compression ratio on these json files to be much better than 1.8x, no?

Reply:
Updated this to 5x of the compressed size, the actual amount will vary, but from docs tends to be between 2:1 and 5:1 so went with the high end of that range.